### ISSUE TITLE
Make scroll to look the default

### ISSUE THREAD

**Issue by JuliusSweetland** [OptiKey/OptiKey#731]
It should absolutely be the default with Free Scrolling and Screen at Point. Deprecate Window at Point, and Window at Center. Deprecate display of Window Bounds. Update the Deadzone bounds to use an optional icon. I have a few examples I made for my AutoHotKey scrolling.

**Comment by JuliusSweetland** [comment]
My thoughts:
* Scroll to look should be the default, with the option, as it is, to turn it off
* I think the full set of settings should be available, but some of the keys we should discuss hiding by default (with a new setting for 'advanced' mode to show them).
* I think there is a case to keep some of the different scroll to look modes (Window At X), but again, perhaps only in the advanced user mode.
* I like the suggestion of an icon reflecting where you are in the scrolling area (if I have understood you correctly), but this implies that we will effectively be turning on magnetic cursor. Can we achieve this without using the cursor and instead display an icon another way so magnetic cursor is still an optional extra you can use/not use alongside everything else.

**Comment by kmcnaught** [comment]
When I've played with look to scroll previously, I've been both overwhelmed by all the options (that really require some technical understanding to grok) and also really impressed by how useful it is. I'd be in favour of a single "type" of look to scroll being the default (e.g. Adam's suggested Free Scrolling + Screen at Point sounds good), with any extra versions being either hidden behind an advanced feature toggle or available only with extra keyboards (we could provide an example keyboard). Even when the advanced options are available we might still want the default (free+point) to be on the main mouse keyboard and the others on a subkeyboard. 

We should try to reduce the complexity somewhat. I count 9 bools in the management console that affect functionality plus 5 types (screen / point / etc) and 4 orientations (horizontal / vertical etc) which gives over 10,000 combinations which is a bit much imho :). But we should discuss the different use cases before dropping too much. 

Some UX thoughts on the different options:
*Invert scrolling direction* - this is a common option for scrolling with a finger action (are you moving the paper with your finger or rotating the viewport?) but I'm not convinced it would ever feel natural to have an inverted direction when you're controlling with eye gaze. I'm happy to be persuaded otherwise. 

*Lock down bounds key...* - should default False since True requires some user understanding of persisted state. Definitely has some use if we're preserving some of the advanced options like "window". I can imagine someone having some documentation open on half their screen, and wanting to return to scrolling the documentation window periodically. However, these users are _probably_ good enough with eye gaze that it doesn't really cost them much to re-click on the documentation window. So I could also be persuaded that it isn't worth the extra state if anyone feels more strongly on this.

*Before scrolling, move mouse to centre of deadzone* - no comment. Not sure if there are corner cases where you would want this to be turned off, similar to #727 ?

*Before scrolling within window, bring it to front* and *After choosing screen point, bring any window there to front*. If I'm understanding these correctly, I'd suggest merging them to a single bool - they do a similar thing in two different scenarios, and I'm not sure why you'd ever want them to be different. Again, I suspect there may be corner cases where this isn't desired, so I'd be happy to keep it as an advanced option, defaulted to True.

*Stop scrolling when [mouse actions]* and *Automatically resume scrolling ...* these are probably fine with Adam's defaults, without exposing them to users. I can't think of *any* situation where you want scrolling and point selection simultaneously so this doesn't even need a setting. We could consider removing "Automatically resume scrolling..." and just having LTS key that can be pressed OR locked down. In the locked case it would resume if locked, but not if only pressed. This might be more consistent with how we usually handle this kind of UX. 

*Stop scrolling when switching keyboards* - I'd suggest it's worth keeping this as an advanced option as I'm sure there are workflows with dynamic keyboards where you don't want this - for instance, consider Adam's extra function keys along the top/left which pop up/down using a ChangeKeyboard key. If your mouse actions are hidden behind this kind of popup functionality, you might not want your scrolling to be turned off each time. If you're using this kind of keyboard you're definitely an advanced enough user to find an option in the MC.

**Types of scrolling**
*Window / Window Region / Custom Region / Screen at point / Screen Centre*
I'm struggling to think of any persuasive user stories for Window Region or Custom Region. The only thing I can think of is if a user wants to set up a semi-permanent scroll bar along the edge of their screen that they leave activated for regular vertical scrolling, for instance. This could be equally achieved with a vertical or cross scroll type, appropriately positioned. 
Window / Screen point / Screen centre seem reasonable options to keep available _somewhere_ (behind advanced feature toggle or on advanced scroll keyboard)

*Free / cross / horizontal /vertical*
Obviously Free captures most use cases. I can picture Cross being easier for someone less confident. I could even envisage an argument for having it as the user-friendly default. Cross / horizontal / vertical could be useful in the "semi-permanent scroll bar" context, so again I think it's reasonable to keep them for advanced use. 

One thought - I know some people might struggle with look to scroll and find discrete actions better. These are likely also the users who wouldn't dive into advanced settings to find "turn on basic scrolling instead". There may be a way we can support both to some extent. For instance (this is just an opening gambit), we could lose "scroll left" / "scroll right" (which I'd expect to be rarely required in a basic workflow) and replace with "look to scroll" and whichever the second-most-important look to scroll key is (speed?). On the rare occasion that beginner user needs to scroll left/right they can have a go with LTS.

**Comment by JuliusSweetland** [comment]
@kmcnaught Good input, thanks. I agree with most things here.

Unfortunately the original contributor Josh Campbell is no longer a GitHub user, but here is the original PR: https://github.com/OptiKey/OptiKey/pull/431

**Comment by kmcnaught** [comment]
That's really helpful context, thanks. I'll read in more detail later.

As an aside, in my own fork I've been playing with having a general-purpose "Interaction2D" class which lets you hook up a different callback to the guts of the LTS engine. This means you can re-use the functionality for e.g. "look to control joystick" or "look up/down to control volume" (I haven't implemented the latter but I think it would be simple). I had to rip out most of the complexity of the scroll options to keep this abstraction tractable (it's only Free + Screen at Point), so it's probably never going to replace LTS completely, but it might be worth considering which parts of the implementation can be made a little more abstract.

This is still in flux in my fork so I'm not intending it to be an immediately concern - more likely these changes we're discussing would happen independently and then I might weigh in with some refactoring to keep it extensible. Just mentioning it to keep in the back of our minds. It would be particularly powerful if it was able to work with the plugin engine.

**Comment by kmcnaught** [comment]
One interesting point raised in the original discussions (including here: https://github.com/OptiKey/OptiKey/issues/423) is that we might make a different call about what we make available on the collapsed keyboard and what's on the expanded keyboard. I've only been looking at expanded since I was just exploring the full gamut of options. 

For instance we might consider always having a "LTS Cross @ Point" key on the collapsed keyboard (which helps less confident people) and "LTS Free @ Point" key on the expanded version (which is most multi-purpose). Then all other "types" of LTS are hidden behind a feature flag / extra keyboard. Just a thought.

**Comment by JuliusSweetland** [comment]
@kmcnaught Agreed, but from my experiments yesterday I quite like LTS Free - I need to test more, but that could be the default from my experience with it.

**Comment by kmcnaught** [comment]
Just to clarify my (possible) concerns with LTS free for a complete beginner / someone struggling with eye tracking. The problem with LTS free is that any movement outside of Optikey causes a scroll event (depending on screen vs window settings). Say Optikey Mouse is at the top of your screen, and you want to scroll down a webpage. You click in the centre of the screen to set the anchor, and then look down until you've reached the right part of the page. Now you want to turn LTS off. You then have to look back to the top of the screen to turn LTS off. If you've got filtering turned up high, or you're just a bit slow / inaccurate heading back up to the Optikey keyboard, then you end up moving through the "scroll up" part of the screen before you manage to turn it off. You could avoid this by placing the centre at the top of the window, but this requires a bit more user strategising.

With LTS cross you have the benefit of only parts of the screen causing scroll events, so you'd have a bit of of a 'safe zone'. 

I'm not strongly arguing for this, I'm completely open to LTS free being the default, but just raising it as a possible concern - we should make sure we test with higher filtering levels and consider this use case along the way

**Comment by JuliusSweetland** [comment]
Valid point. If we keep cross we'd also need the boundary lines to be drawn
as otherwise it's difficult to understand where you can trigger scrolling.
Adam advocated for removing the boundary lines I believe.

**Comment by AdamRoden** [comment]
Here's a solution, when the LTS is anchored add a low distraction scroll icon to the deadzone and make it tie back to the LTS key, so the user can look at the icon to pause or stop scrolling.

One idea for an icon is a transparent circle outlined with black and white on the edge that, when you scroll, shows an indicator toward your gaze is in relation to the cursor. A sexy way to do it would be to distort the circle with a parabola toward your gaze.

**Comment by kmcnaught** [comment]
Yeah, I do like the idea of having a "stop key" at the deadzone, though I'm not sure if it would always give the right kinds of UX. I'm thinking that if you're using the scroll to read a long-form article, you probably want to be able to scroll down a bit, then read all the text in the middle, then scroll down a bit more, etc. I'd worry an icon and dwell response would be distracting, and you could find yourself turning it off inadvertently. 

I guess there are two different types of scroll functionality people might want - one is the option to do a single scroll action, but with precise control (and the opportunity to look around a bit while doing this) - for instance, scroll down to the news headline you want to click on. In this case being able to disable from the deadzone is ideal. Separately, there are scenarios in which you want continuous "browsing" while scrolling, where you might scroll for a while, read some text, and then resume scrolling. 

I'm not sure how best to support both in the simplest way...

**Comment by JuliusSweetland** [comment]
@kmcnaught @AdamRoden When I was experimenting by reading websites I found the free scroll worked very well to follow where I was reading. Text scrolled to the centre, where I read it, and then the page followed me as I looked further down.

**Comment by AdamRoden** [comment]
@JuliusSweetland @kmcnaught So, who is taking point and going to do the bulk of this? I'm no longer focusing on smoothing, so I could start a new branch with these changes unless one of you feel like doing it.

My vote for default scroll method is Free. Another thought is that before scrolling based on point position, we check the point. If the point is on a key or if the delta between the point and the previous point is large we do not scroll.

**Comment by JuliusSweetland** [comment]
If you're happy to drive that would be great. I think free as well, and definitely no scrolling if on a key. No scrolling when large siccades sounds sensible as well.

Shall we get a list of changes in here so we can work it out before any coding? For example are any of the look to scroll options going to be on the expanded mouse keyboard? Where should the manual click to scroll keys be - hidden behind a setting somewhere? Which look to scroll modes should survive? Etc

**Comment by AdamRoden** [comment]
Do you want to edit your top comment to include a running list of requirements?  You could get rid of my pics. They're not needed.

**Comment by JuliusSweetland** [comment]
Sure, but I'll have to do it in the morning.

**Comment by JuliusSweetland** [comment]
@AdamRoden @kmcnaught I've updated the top comment. I don't want to distract from finalising the smoothing work too much, but we can start collecting ideas here if anyone is interested in starting to think about this.

**Comment by kmcnaught** [comment]
FYI I'm only engaging in the discussion here, I don't plan to implement any of this - if Adam's got the time to contribute, that's great by me! Thanks for the updated summary Julius. I think I've already covered my current position on most of the points, but I'm keen to hear other people's perspectives / rebuttals.

I'm happy with aiming for Free to be default so long as we make sure we're testing/reviewing with a "really unconfident user with maximum smoothing" in mind. I suspect this isn't going to be a problem. 

Was the suggestion for adaptive icons meant to be an icon at your gaze point or an icon at the deadzone? I thought Adam was implying it would be at your deadzone / reference point, but in that case you wouldn't be seeing it while scrolling. I do like the idea of having some kind of signifier so you're conscious that you're still in a scrolling state, but it would need to be quite unobtrusive. 

I like the suggestion of disabling scroll actions for saccades.

**Comment by JuliusSweetland** [comment]
For the easy win I'm not against the guide rails (the yellow lines) showing
you that you're in a look to scroll state, and where the boundaries are.
They may not be intuitive, but once you know what they are they certainly
do the job.

**Comment by kmcnaught** [comment]
Re the boundary lines, there are a couple of different parts and we might be talking about different bits at different times.

1) The deadzone. This is either a coloured rectangle (free mode), or horizontal and/or vertical "guide rails" (horz/vert/cross mode). It isn't particularly pretty but it works for giving the required feedback. I think it's definitely important that there is some signifier for the centre, and some signifier for the type of scrolling (e.g. free vs horizontal). Having actual deadzone bounds can also be helpful but probably isn't critical if anyone doesn't like them. So we could make it prettier with an icon and/or a more subtle deadzone or whatever but I'm also not against keeping it as-is for now.

2) The bounds. This is the rectangular box around the screen or window or area that you're scrolling in. I don't think it's necessary for "screen" mode, but if we're keeping any of the other advanced options (e.g. window, which I think there's an argument for keeping behind an advanced feature toggle) it is helpful here. It might be that we just turn it off for the "screen" mode(s) to reduce visual clutter. Also, it's possible we can tweak the defaults so it's a bit nicer-looking - maybe thicker but with transparency.

I like the thought of changing the windows icon. Are there any situations where moving the mouse out of some UI bounds loses focus or prevents the scroll events from being sent to the correct UI element? It feels plausible but I can't think of an example.

**Comment by kmcnaught** [comment]
@JuliusSweetland just had a play and adding in magnetic cursor would definitely be a problem, unless we changed the scrolling functionality so the scroll button gets held down throughout the process.

At the moment, the mouse cursor remains in the location you've picked, which (in "screen" mode) means you can use the whole screen to control the scroll, but the `mouseOutputService.ScrollWheelAbsolute` actions get sent to whatever is at the cursor location. This means you can start the scroll action inside a scrollable UI element (like this small comment edit box I'm typing into) and wherever you look on screen, the scrolling will only occur in the edit box. If you moved the cursor too, the scroll events would be sent to the parent webpage as soon as your gaze left the edit box, which would make it very hard to scroll within a small scrollable region.

It's possible we could turn on magnetic cursor _and_ keep this behaviour if we sent a "mouse wheel held down" signal (is this equivalent to middle button?), but I think the mouse movements would cause scrolling natively and prevent us from controlling the speed.

**Comment by JuliusSweetland** [comment]
Very good spot. I think keeping the cursor in one spot is probably the
correct behaviour. What is the experience like if you turn on magnetic
cursor and then try look to scroll? Is it difficult to understand what is
happening/wrong behaviour?

**Comment by kmcnaught** [comment]
With magnetic cursor + LTS, the area in which the scrolling is applied changes once your gaze leaves the original area. I can see this behaviour being understandable / potentially useful in some simple scenarios (like the second example below) but it would cause significant problems when you're trying to scroll a small scrollable element (like the first). I think the downside of that failure mode significantly outweighs any upside in easier situations. 

Trying to scroll small comment box (this is great without mag cursor and almost impossible with):
https://www.loom.com/share/6659fc634d414d718ebd44d158a11a3b

Scrolling large embedded code block (this behaviour doesn't look so bad):
https://www.loom.com/share/9179445cd02a46fabb2fb636d01e2e03

**Comment by JuliusSweetland** [comment]
Useful videos. As the user has the option of combining magnetic cursor with
lts my vote is for keeping it as it is, i.e. static cursor by default. If
we do this there's nothing stopping us from changing the cursor icon to
reflect our scrolling actions, it just wouldn't follow your gaze. Would
this be a major problem?

**Comment by kmcnaught** [comment]
That sounds fine by me. I think @AdamRoden had been suggesting adaptive icons showing the scroll effect, which I can only see being helpful _if_ the icon is at your gaze point, but I personally don't think it's necessary to see the current effect so long as there's something (icon or guidelines) showing the type of effect (horz/vert/etc). It's also possible I misunderstood, so @AdamRoden any thoughts?

**Comment by AdamRoden** [comment]
Ideally, when you place the LTS anchor it should move the cursor there, bring that window to the front, pause the magnetic cursor, and instead display either a fixed position key that updates appearance based on gaze position, or a popup icon that follows your gaze and updates appearance based on distance and angle from the anchored cursor.

**Comment by JuliusSweetland** [comment]
@AdamRoden @kmcnaught Any potential downsides of pausing magnetic cursor when using LTS? I can't think of a situation where I'd want both - you guys?

**Comment by kmcnaught** [comment]
Off the top of my head - browsing something which has mouseover tooltips. I don't know how often you'd be in a situation where you want to do that within a scrollable area. What's the implied scenario where you would be using magnetic cursor beforehand but *not* want it while scrolling? They seem fairly independent choices to me. @AdamRoden what are the most common scenarios you use magnetic cursor for?

**Comment by AdamRoden** [comment]
Here's where I'm at for the settings and keyboards. The biggest change is that I removed LTS Bounds keys and made it so that option is only available in MC. Beyond that, I eliminated a few other options. I haven't done anything fancy, and I don't plan to at this juncture. Instead, let's leave that as a separate improvement and stick to discussing the necessities for now.
How do these screenshots look? I don't like the empty space, but it's needed for the old fashioned look to scroll if we are keeping that as an option. 

![image](https://user-images.githubusercontent.com/45572061/95518866-96329900-0989-11eb-922b-12a7be54f9cf.png)
![image](https://user-images.githubusercontent.com/45572061/95518940-ba8e7580-0989-11eb-8734-cb6c25a5228a.png)
![image](https://user-images.githubusercontent.com/45572061/95518971-cd08af00-0989-11eb-8e36-5908e509b732.png)
![image](https://user-images.githubusercontent.com/45572061/95519005-de51bb80-0989-11eb-83bd-9242ba282e0b.png)

**Comment by JuliusSweetland** [comment]
Looking good. We can shuffle the keys around later to make everything look nice. Couple of thoughts:

1.If possible we should aim to separate manual scrolling keys from the LTS keys - as you've started doing we could do this by keeping them on different rows. I'm not sure how we could make this more obvious, but just looking at the keys even I have to go back and remind myself what they all do.
2.Are we sure we want to keep the free/cross key? I would be happy to just default to free, with an MC setting for other types. I think it's a one time decision, rather than something the user would want to change regularly.

**Comment by AdamRoden** [comment]
Here are my personal thoughts. The keys from the mouse keyboard should all be merged onto the web browsing keyboard and be renamed to Browsing, Navigating, or Exploring. Then, the new Browsing keyboard needs to adopt the same common keys found on the letters, numbers, and currencies keyboards, and maybe even keep the scratchpad for continuity. Finally, get rid of the full-size mouse keyboard and make it always open the collapsed version.

![image](https://user-images.githubusercontent.com/45572061/95595812-37653200-0a12-11eb-950f-620513bb5a12.png)

**Comment by JuliusSweetland** [comment]
I could certainly get on board with enriching the browsing keyboard by
adding some mouse keys and possibly some other keys, but I'm less keen on
losing the expanded mouse keyboard.

**Comment by AdamRoden** [comment]
Mouse and Browsing all-in-one. Redundant otherwise. Don't you think?

**Comment by JuliusSweetland** [comment]
I get where you're coming from, but 2 things stick for me:

1.I don't think the browser keyboard needs *all* mouse keys. E.g. move to point.
2.There's a stand alone Optikey Mouse application, which doesn't need the browsing components, so both need to exist in some form anyway.

I do agree that the browser keyboard can be improved. It's good for what it does, but it requires you to move back and forth between keyboard to get anything done in a browser.

**Comment by AdamRoden** [comment]
Can I attempt to make the keyboards provide both the manual scroll keys and a singular LTS key, in such a way that it is not an either/or decision? All LTS configuration options would be in MC only.

**Comment by JuliusSweetland** [comment]
Yes that sounds ideal. Scroll keys (both kinds) could then be added to the
browser keyboard if you agree that there's value doing that.

**Comment by JuliusSweetland** [comment]
@AdamRoden Are you still planning to improve the LTS experience? Just working out what is still relevant here.

**Comment by AdamRoden** [comment]
@JuliusSweetland I thought we were satisfied with what I had done. There is one undesirable I should address: if you glance off-screen and then return it sometimes scrolls by a large amount. Why don't I fix that and then we can close this ticket.

**Comment by JuliusSweetland** [comment]
Ideal yes. I wasn't raising any concerns, just wondering if you still
wanted to use icons or anything else. I'm very happy with how it is now.
